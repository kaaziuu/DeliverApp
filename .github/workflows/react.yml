name: react

on: [push, pull_request]

jobs:
    build:
        runs-on: ubuntu-latest
    steps:
        - uses: actions/checkout@v3
        - name: Setup node
          uses: actions/setup-node@v3
          with:
              note-version: 16.x
              cache: "npm"
      - run: npm  install .\jsApps\launcher\
      - run: npm --prefix .\jsApps\launcher\ run test

