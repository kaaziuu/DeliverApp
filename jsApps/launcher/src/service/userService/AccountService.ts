import axios, { AxiosError, AxiosResponse } from "axios";
import { useMutation, useQuery } from "react-query";
import { HandleApiError } from "service/_core/HandleApiError";
import { BaseResponse, FetchProcessing, MutationProcessing } from "service/_core/Models";
import { UseStore } from "stores/Store";
import Endpoints from "utils/axios/Endpoints";
import GetHeader from "utils/axios/GetHeader";
import Config from "utils/_core/Config";
import ChangePasswordForm from "./models/AccountModels/ChangePasswordForm";

const ChangePasswordRequest = async (request: ChangePasswordForm, jwt: string): Promise<BaseResponse<null>> => {
    const header = GetHeader(jwt);

    const response = await axios
        .put<ChangePasswordForm, AxiosResponse<BaseResponse<null>>>(
            `${Config.serverUrl}${Endpoints.Account.ChagnePassword}`,
            request,
            { headers: header }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError(error));

    return response;
};

export const ChangePasswordAction = (): MutationProcessing<ChangePasswordForm, BaseResponse<null>> => {
    const { userStore } = UseStore();

    const { isLoading, data, mutateAsync } = useMutation((form: ChangePasswordForm) =>
        ChangePasswordRequest(form, userStore.getUser!.jwt)
    );

    return {
        isLoading: isLoading,
        mutateAsync: mutateAsync,
        data: data,
        error: data?.error,
        isSuccess: data?.isSuccess,
    };
};

const IsValidRecoveryKeyRequest = async (key: string): Promise<BaseResponse<Boolean>> => {
    const reponse = await axios
        .get<null, AxiosResponse<BaseResponse<Boolean>>>(
            `${Config.serverUrl}${Endpoints.Account.PasswordRecoverySetNewPassword}`
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError<Boolean>(error));

    return reponse;
};

export const IsValidRecoveryKeyAction = (recoveryKey: string): FetchProcessing<Boolean> => {
    const { isLoading, data, refetch } = useQuery("IsValidRecoveryKeyRequest", () =>
        IsValidRecoveryKeyRequest(recoveryKey)
    );
    return {
        isLoading: isLoading,
        data: data?.data,
        isSuccess: data?.isSuccess,
        error: data?.error,
        refresh: refetch,
    };
};
