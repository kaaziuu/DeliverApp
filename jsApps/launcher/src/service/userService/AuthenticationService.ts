import axios, { AxiosError, AxiosResponse } from "axios";
import { useMutation, useQuery } from "react-query";
import { useNavigate } from "react-router-dom";
import GetHeader from "utils/axios/GetHeader";
import Path from "utils/route/Path";
import { UseStore } from "../../stores/Store";
import Endpoints from "../../utils/axios/Endpoints";
import Config from "../../utils/_core/Config";
import { BaseResponse, FetchProcessing, MutationProcessing } from "../_core/Models";
import AuthResponse from "./models/AuthResponse";
import LoginForm from "./models/LoginForm";

export const LoginRequest = async (form: LoginForm): Promise<BaseResponse<AuthResponse>> => {
    const response = await axios
        .post<LoginForm, AxiosResponse<BaseResponse<AuthResponse>>>(
            `${Config.serverUrl}${Endpoints.Authentication.Login}`,
            form,
            {
                withCredentials: true,
            }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => {
            return {
                isSuccess: false,
                error: error.message,
            } as BaseResponse<AuthResponse>;
        });
    return response;
};

const RefreshTokenRequest = async (): Promise<BaseResponse<AuthResponse>> => {
    const response = await axios
        .post<undefined, AxiosResponse<BaseResponse<AuthResponse>>>(
            `${Config.serverUrl}${Endpoints.Authentication.Refresh}`,
            null,
            {
                withCredentials: true,
            }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => {
            return {
                isSuccess: false,
                error: error.message,
            } as BaseResponse<AuthResponse>;
        });
    return response;
};

export const RefreshToken = (): FetchProcessing<AuthResponse> => {
    const { userStore } = UseStore();
    const { isLoading, data } = useQuery("RefreshToken", RefreshTokenRequest, {
        refetchInterval: 180000,
        onError: () => {
            userStore.logout();
        },
        onSuccess: (response) => {
            if (!response.isSuccess) {
                userStore.logout();
            }
        },
    });

    if (!isLoading && data?.isSuccess) {
        userStore.setUser(data!.data!);
    }

    return {
        isLoading: isLoading,
        error: data?.error,
        isSuccess: data?.isSuccess,
        data: data?.data,
    };
};

export const Login = (): MutationProcessing<LoginForm, BaseResponse<AuthResponse>> => {
    const { userStore } = UseStore();
    const navigation = useNavigate();

    const { isLoading, data, mutate, mutateAsync } = useMutation((request: LoginForm) => LoginRequest(request), {
        onSuccess: (result) => {
            if (result?.isSuccess) {
                userStore.setUser(result!.data!);
            } else {
                navigation(Path.login);
            }
        },
    });

    return {
        isLoading: isLoading,
        error: data?.error,
        isSuccess: data?.isSuccess,
        data: data,
        mutateAsync: mutateAsync,
    };
};

const LogoutRequest = async (jwt: string): Promise<BaseResponse<null>> => {
    const reponse = await axios
        .post<null, AxiosResponse<BaseResponse<null>>>(
            `${Config.serverUrl}${Endpoints.Authentication.Logout}`,
            {},
            { headers: GetHeader(jwt) }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => ({ error: error.message, isSuccess: false } as BaseResponse<null>));

    return reponse;
};

export const Logout = (): MutationProcessing<undefined, BaseResponse<null>> => {
    const { userStore } = UseStore();

    const { isLoading, data, mutate, mutateAsync } = useMutation("logout", () => LogoutRequest(userStore.getUser!.jwt));

    return {
        isLoading: isLoading,
        data: data,
        error: data?.error,
        isSuccess: data?.isSuccess,
        mutateAsync: mutateAsync,
    };
};
