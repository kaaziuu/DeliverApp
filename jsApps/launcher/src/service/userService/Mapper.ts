import AuthResponse from "./models/AuthModels/AuthResponse";
import User from "./models/User";

export const MapAuthResponseToUser = (response: AuthResponse): User => {
    return {
        hash: response.hash,
        jwt: response.jwt,
        name: response.name,
        roles: response.roles,
        surname: response.surname,
        username: response.surname,
        expireDate: response.expireDate,
        companyHash: response.companyHash,
    };
};

export default MapAuthResponseToUser;
