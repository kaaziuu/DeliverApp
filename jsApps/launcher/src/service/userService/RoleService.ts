import axios, { AxiosError, AxiosRequestHeaders, AxiosResponse } from "axios";
import { useQuery } from "react-query";
import { BaseResponse, FetchProcessing } from "service/_core/Models";
import { UseStore } from "stores/Store";
import Endpoints from "utils/axios/Endpoints";
import GetHeader from "utils/axios/GetHeader";
import Config from "utils/_core/Config";
import RoleResponse from "./models/RoleResponse";

const FetchRole = async (header: AxiosRequestHeaders): Promise<BaseResponse<RoleResponse[]>> => {
    const response = await axios
        .get<null, AxiosResponse<BaseResponse<RoleResponse[]>>>(`${Config.serverUrl}${Endpoints.Role.GetAll}`, {
            headers: header,
        })
        .then((resp) => resp.data)
        .catch(
            (error: AxiosError) =>
                ({
                    isSuccess: false,
                    error: error.message,
                } as BaseResponse<RoleResponse[]>)
        );

    return response;
};

export const GetRoles = (): FetchProcessing<RoleResponse[]> => {
    const { userStore } = UseStore();
    const header = GetHeader(userStore.getUser!.jwt);
    const { data, isLoading } = useQuery("fetch roles", () => FetchRole(header), { cacheTime: Infinity });

    return {
        isLoading: isLoading,
        data: data?.data,
        error: data?.error,
        isSuccess: data?.isSuccess,
    };
};
