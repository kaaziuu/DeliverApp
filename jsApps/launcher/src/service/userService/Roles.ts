export const Roles = {
    Admin: "Admin",
    CompanyAdmin: "CompanyAdmin",
    CompanyOwner: "CompanyOwner",
    Hr: "HR",
    Driver: "Driver",
    Dispatcher: "Dispatcher",
};

const HasRole = (roles: string[], requiresRole: string[]): boolean => {
    const hasRole = roles.some((x) => requiresRole.some((y) => y === x));
    return hasRole;
};

export default HasRole;
