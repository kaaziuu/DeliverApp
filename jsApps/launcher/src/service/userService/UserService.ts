import axios, { AxiosError, AxiosRequestHeaders, AxiosResponse } from "axios";
import { useMutation, useQuery } from "react-query";
import { HandleApiError } from "service/_core/HandleApiError";
import { UseStore } from "stores/Store";
import Endpoints from "utils/axios/Endpoints";
import GetHeader from "utils/axios/GetHeader";
import Config from "utils/_core/Config";
import { BaseResponse, FetchProcessing, MutationProcessing } from "../_core/Models";
import CreateUserForm from "./models/UserModels/CreateUserForm";
import UpdateUserForm from "./models/UserModels/UpdateUserForm";
import UserResponse from "./models/UserModels/UserResponse";

const CreateUserRequest = async (
    form: CreateUserForm,
    header: AxiosRequestHeaders
): Promise<BaseResponse<UserResponse>> => {
    const response = await axios
        .post<CreateUserForm, AxiosResponse<BaseResponse<UserResponse>>>(
            `${Config.serverUrl}${Endpoints.User.Create}`,
            form,
            { headers: header }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError<UserResponse>(error));
    return response;
};

export const CreateUser = (): MutationProcessing<CreateUserForm, BaseResponse<UserResponse>> => {
    const { userStore } = UseStore();
    const header = GetHeader(userStore.getUser!.jwt);

    const { isLoading, mutateAsync } = useMutation((form: CreateUserForm) => CreateUserRequest(form, header), {
        onMutate: (form) => {
            form.companyHash = userStore.getUser!.companyHash;
        },
    });

    return {
        isLoading: isLoading,
        mutateAsync: mutateAsync,
    };
};

const GetUserRequest = async (hash: string, jwt: string): Promise<BaseResponse<UserResponse>> => {
    const header = GetHeader(jwt);

    const response = await axios
        .get<null, AxiosResponse<BaseResponse<UserResponse>>>(`${Config.serverUrl}${Endpoints.User.GetUser(hash)}`, {
            headers: header,
        })
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError<UserResponse>(error));

    return response;
};

export const GetUser = (
    hash: string,
    onSuccess?: (result: BaseResponse<UserResponse>) => void
): FetchProcessing<UserResponse> => {
    const { userStore } = UseStore();

    const { isLoading, data } = useQuery(["GetUser", hash], () => GetUserRequest(hash, userStore.getUser!.jwt), {
        cacheTime: Infinity,
        onSuccess: onSuccess,
    });

    return {
        isLoading: isLoading,
        isSuccess: data?.isSuccess,
        data: data?.data,
        error: data?.error,
    };
};

const UpdateUserRequest = async (request: UpdateUserForm, jwt: string): Promise<BaseResponse<UserResponse>> => {
    const header = GetHeader(jwt);

    const response = await axios
        .put<UpdateUserForm, AxiosResponse<BaseResponse<UserResponse>>>(
            `${Config.serverUrl}${Endpoints.User.UpdateUser}`,
            request,
            {
                headers: header,
            }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError<UserResponse>(error));

    return response;
};

export const UpdateUserAction = (): MutationProcessing<UpdateUserForm, BaseResponse<UserResponse>> => {
    const { userStore } = UseStore();

    const { isLoading, mutateAsync } = useMutation("update user", (form: UpdateUserForm) =>
        UpdateUserRequest(form, userStore.getUser!.jwt)
    );

    return {
        isLoading: isLoading,
        mutateAsync: mutateAsync,
    };
};

const FetchWorkers = async (header: AxiosRequestHeaders): Promise<BaseResponse<UserResponse[]>> => {
    const response = await axios
        .get<BaseResponse<UserResponse[]>>(`${Config.serverUrl}${Endpoints.User.UserList}`, {
            headers: header,
        })
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError<UserResponse[]>(error));
    return response;
};

export const GetWorkers = (): FetchProcessing<UserResponse[]> => {
    const { userStore } = UseStore();

    const header = GetHeader(userStore.getUser!.jwt);
    const { data, isLoading, refetch } = useQuery("fetch workers", () => FetchWorkers(header), { cacheTime: Infinity });

    return {
        isLoading: isLoading,
        data: data?.data,
        error: data?.error,
        isSuccess: data?.isSuccess,
        refresh: refetch,
    };
};

const FireUserRequest = async (userHash: string, jwt: string): Promise<BaseResponse<null>> => {
    const response = await axios
        .put<null, AxiosResponse<BaseResponse<null>>>(
            `${Config.serverUrl}${Endpoints.User.Fire(userHash)}`,
            undefined,
            { headers: GetHeader(jwt) }
        )
        .then((resp) => resp.data)
        .catch((error: AxiosError) => HandleApiError(error));

    return response;
};

export const FireUserAction = (): MutationProcessing<string, BaseResponse<null>> => {
    const { userStore } = UseStore();

    const { isLoading, mutateAsync } = useMutation((hash: string) =>
        FireUserRequest(hash, userStore.getUser?.jwt ?? "")
    );

    return {
        isLoading: isLoading,
        mutateAsync: mutateAsync,
    };
};
