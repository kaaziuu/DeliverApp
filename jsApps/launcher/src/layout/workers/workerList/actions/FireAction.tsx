import PersonRemoveIcon from "@mui/icons-material/PersonRemove";
import { LoadingButton } from "@mui/lab";
import { Portal } from "@mui/material";
import Snackbar from "components/snackbar/Snackbar";
import { FC, useState } from "react";
import { FireUserAction, GetWorkers } from "service/userService/UserService";

interface FireActionProps {
    userHash: string;
}

const FireAction: FC<FireActionProps> = ({ userHash }) => {
    const { isLoading, mutateAsync } = FireUserAction();
    const [error, setError] = useState<string | null>(null);
    const [showSnackbar, setShowSnackbar] = useState(false);
    const { refresh } = GetWorkers();

    const submit = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
        e.stopPropagation();
        const result = await mutateAsync(userHash);
        if (result.isSuccess) {
            setError(null);
        } else {
            setError(result.error);
        }
        setShowSnackbar(true);
        refresh!();
    };

    return (
        <>
            <LoadingButton loading={isLoading} onClick={(e) => submit(e)}>
                <PersonRemoveIcon />
            </LoadingButton>
            <Portal container={document.querySelector("#root")}>
                <Snackbar
                    message={error ?? "Worker was fired"}
                    setIsOpen={setShowSnackbar}
                    variant={error ? "error" : "success"}
                    isOpen={showSnackbar}
                />
            </Portal>
        </>
    );
};

export default FireAction;
