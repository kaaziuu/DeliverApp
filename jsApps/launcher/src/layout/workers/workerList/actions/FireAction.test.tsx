import { render } from "@testing-library/react";
import { FireUserAction } from "service/userService/UserService";
import FireAction from "./FireAction";

jest.mock("service/userService/UserService", () => ({
    FireUserAction: jest.fn(),
}));

describe("FireAction", () => {
    test("should render Fire action button", () => {
        // arrnage
        (FireUserAction as jest.MockedFunction<typeof FireUserAction>).mockReturnValue({
            isLoading: false,
            mutateAsync: jest.fn(),
        });

        // act
        const component = render(<FireAction userHash="hash" />);

        // assert
        expect(component).toMatchSnapshot();
    });
});
