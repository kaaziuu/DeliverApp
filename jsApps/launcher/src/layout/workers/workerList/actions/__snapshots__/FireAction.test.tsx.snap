// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FireAction should render Fire action button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="MuiLoadingButton-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1vqlznj-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
        id="mui-1"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="PersonRemoveIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M14 8c0-2.21-1.79-4-4-4S6 5.79 6 8s1.79 4 4 4 4-1.79 4-4zm3 2v2h6v-2h-6zM2 18v2h16v-2c0-2.66-5.33-4-8-4s-8 1.34-8 4z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="MuiLoadingButton-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root css-1vqlznj-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
      id="mui-1"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="PersonRemoveIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M14 8c0-2.21-1.79-4-4-4S6 5.79 6 8s1.79 4 4 4 4-1.79 4-4zm3 2v2h6v-2h-6zM2 18v2h16v-2c0-2.66-5.33-4-8-4s-8 1.34-8 4z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
